Sources/TuistGenerator/Linter/StaticProductsGraphLinter.swift

```diff
@@ -0,0 +1,166 @@
+import Basic
+import Foundation
+import TuistCore
+
+protocol StaticProductsGraphLinting {
+    func lint(graph: Graphing) -> [LintingIssue]
+}
+
+class StaticProductsGraphLinter: StaticProductsGraphLinting {
+    func lint(graph: Graphing) -> [LintingIssue] {
+        let nodes = graph.entryNodes
+        return warnings(in: nodes)
+            .sorted()
+            .map(lintIssue)
+    }
+
+    private func warnings(in nodes: [GraphNode]) -> Set<StaticDependencyWarning> {
+        var warnings = Set<StaticDependencyWarning>()
+        let cache = Cache()
+        nodes.forEach { node in
+            // Skip already evaluated nodes
+            guard cache.results(for: node) == nil else {
+                return
+            }
+            let results = buildStaticProductsMap(visiting: node,
+                                                 cache: cache)
+            warnings.formUnion(results.linked.flatMap(staticDependencyWarning))
+        }
+        return warnings
+    }
+
+    private func staticDependencyWarning(staticProduct: GraphNode, linkedBy: Set<TargetNode>) -> [StaticDependencyWarning] {
+        // Common dependencies between test bundles and their host apps are automatically omitted
+        // during generation - as such those shouldn't be flagged
+        //
+        // reference: https://github.com/tuist/tuist/pull/664
+        let apps: Set<GraphNode> = linkedBy.filter { $0.target.product == .app }
+        let hostedTestBundles = linkedBy
+            .filter { $0.target.product.testsBundle }
+            .filter { $0.dependencies.contains(where: { apps.contains($0) }) }
+
+        let links = linkedBy.subtracting(hostedTestBundles)
+
+        guard links.count > 1 else {
+            return []
+        }
+
+        let sortedLinks = links.sorted(by: { $0.name < $1.name })
+        return [
+            .init(staticProduct: staticProduct,
+                  linkingNodes: sortedLinks),
+        ]
+    }
+
+    private func buildStaticProductsMap(visiting node: GraphNode,
```

Comment:
Maybe it is worth explaining this briefly in a comment

Reply:
Good point added an explanation on what this does - hope it makes sense